File: android/app/src/main/java/com/odelly/pypet/PypetCurriculum.java
package com.odelly.pypet;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/** Full-spectrum Python-to-job-readiness curriculum used by the in-world Academy. */
public final class PypetCurriculum {
    public static final class Lesson {
        public final String id, title, skill, task;
        public Lesson(String id, String title, String skill, String task) { this.id=id; this.title=title; this.skill=skill; this.task=task; }
    }
    public static final class Module {
        public final String title, description;
        public final List<Lesson> lessons;
        Module(String title, String description, Lesson... lessons) { this.title=title; this.description=description; this.lessons=Collections.unmodifiableList(Arrays.asList(lessons)); }
    }
    private static Lesson l(String id,String title,String skill,String task){return new Lesson(id,title,skill,task);}
    public static final List<Module> MODULES = Collections.unmodifiableList(Arrays.asList(
        new Module("01 • Foundations","Learn Python by changing real things in the World.",
            l("f01","Python & the REPL","syntax","Print a greeting for your pet."),l("f02","Variables & types","variables","Store your pet's name, age and happiness."),l("f03","Numbers & operators","arithmetic","Calculate food and coin totals."),l("f04","Strings","strings","Format a personalized pet message."),l("f05","Booleans & logic","logic","Decide whether your pet needs care."),l("f06","Input & output","io","Ask the player for a town name."),l("f07","if / elif / else","conditionals","Choose an action from pet needs."),l("f08","for loops","loops","Process every item in a supply list."),l("f09","while loops","loops","Run a safe care loop until a need is satisfied.")),
        new Module("02 • Core Python","Build reusable programs and work with collections.",
            l("c01","Functions","functions","Write a feed_pet function with parameters and a return value."),l("c02","Lists & tuples","collections","Track a party of pets."),l("c03","Dictionaries & sets","collections","Build a pet profile and remove duplicate skills."),l("c04","Comprehensions","collections","Create a filtered list of healthy pets."),l("c05","Exceptions","errors","Handle invalid player input without crashing."),l("c06","Files","files","Save and reload a town note."),l("c07","Modules & imports","modules","Split a pet utility into an importable module."),l("c08","Packages & environments","packaging","Create a small package with a clean entry point.")),
        new Module("03 • Standard Library","Learn the practical Python library used in real work.",
            l("s01","pathlib","stdlib","Find and create files safely."),l("s02","json","stdlib","Save a pet profile as JSON."),l("s03","csv","stdlib","Export a pet-care report."),l("s04","datetime","stdlib","Calculate streak dates."),l("s05","math & statistics","stdlib","Analyze training results."),l("s06","random","stdlib","Create fair non-cheating game events."),l("s07","re","stdlib","Validate structured text."),l("s08","collections","stdlib","Use Counter, defaultdict and deque."),l("s09","itertools & functools","stdlib","Compose reusable data pipelines."),l("s10","logging","stdlib","Add useful application diagnostics.")),
        new Module("04 • Professional Python","Write maintainable software.",
            l("p01","Classes & objects","oop","Model a Pet class with state and behavior."),l("p02","Inheritance & composition","oop","Build species-specific behavior without duplication."),l("p03","Dataclasses","oop","Represent a typed pet record."),l("p04","Decorators","advanced","Add timing or validation around a function."),l("p05","Generators","advanced","Stream a large training dataset."),l("p06","Context managers","advanced","Safely manage a resource."),l("p07","Type hints","typing","Annotate a real module."),l("p08","Testing","testing","Write unit tests for pet-care rules."),l("p09","Debugging & profiling","debugging","Find and fix a deliberate bug and slow path."),l("p10","Packaging & dependency management","packaging","Prepare a professional installable project.")),
        new Module("05 • Data, Web & Systems","Build applications that communicate and persist data.",
            l("w01","HTTP & APIs","web","Call a safe sample API and parse JSON."),l("w02","SQLite & SQL","database","Create and query a pet-care database."),l("w03","Data processing","data","Transform and summarize records."),l("w04","CLI applications","cli","Build a useful command-line pet tool."),l("w05","Concurrency","systems","Compare threads, processes and async tasks."),l("w06","asyncio","systems","Run asynchronous work safely."),l("w07","Security fundamentals","security","Validate input and protect secrets.")),
        new Module("06 • Tkinter & GUI","Create desktop interfaces with events and state.",
            l("t01","Tkinter windows","tkinter","Build a working pet-care window."),l("t02","Widgets & layout","tkinter","Arrange labels, buttons, entries and frames."),l("t03","Events & callbacks","tkinter","Connect UI actions to pet behavior."),l("t04","Forms & validation","tkinter","Build a validated profile form."),l("t05","GUI project","tkinter","Ship a small multi-screen desktop app.")),
        new Module("07 • Pygame & 2D Graphics","Learn game loops, input, collision, sound and animation.",
            l("g01","Game loop","pygame","Build an update/draw loop."),l("g02","Coordinates & resolutions","graphics","Map points correctly at multiple resolutions and aspect ratios."),l("g03","Sprites","pygame","Create and move a player sprite."),l("g04","Input","pygame","Handle keyboard/controller input."),l("g05","Collision","pygame","Implement safe collision detection."),l("g06","Animation","pygame","Build an idle and walking animation."),l("g07","Sound","pygame","Trigger toggleable SFX and music."),l("g08","Camera & scaling","graphics","Build a camera that follows a pet."),l("g09","2D capstone","pygame","Ship a small playable game.")),
        new Module("08 • 3D Graphics & Animation","Understand the math and design behind immersive worlds.",
            l("d01","3D coordinates","3d","Map x/y/z positions and camera space."),l("d02","Vectors","3d","Move a character using vectors."),l("d03","Transforms","3d","Translate, rotate and scale an object."),l("d04","Meshes","3d","Understand vertices, edges and faces."),l("d05","Materials & lighting","3d","Create readable surfaces and lighting."),l("d06","Camera systems","3d","Build perspective and follow-camera behavior."),l("d07","3D animation","3d","Create a simple character animation state machine."),l("d08","Optimization","3d","Measure and improve frame performance."),l("d09","3D capstone","3d","Build and present a small interactive scene.")),
        new Module("09 • Job Readiness & Mastery","Turn skills into demonstrable professional ability.",
            l("j01","Git & GitHub","career","Create clean commits and a useful README."),l("j02","Code review","career","Review a change and give actionable feedback."),l("j03","Documentation","career","Document an API and setup process."),l("j04","Architecture","career","Design modules with clear responsibilities."),l("j05","Portfolio project","career","Build a complete Python application."),l("j06","Technical interview","career","Solve and explain coding problems."),l("j07","Debugging interview","career","Diagnose a broken program methodically."),l("j08","Presentation","career","Explain your project and design decisions."),l("j09","Master capstone","career","Plan, build, test and present a professional project."))
    ));
    private PypetCurriculum() {}
    public static int lessonCount(){int n=0;for(Module m:MODULES)n+=m.lessons.size();return n;}
}

File: android/app/src/main/python/pypet_curriculum.py
"""PyPet Academy: hands-on novice-to-master Python curriculum."""
from dataclasses import dataclass, asdict
from typing import Any

@dataclass(frozen=True)
class Lesson:
    id: str
    tier: str
    title: str
    skill: str
    mission: str
    module: str = "python"
    world_effect: str = "knowledge"

TIERS = ("Novice", "Apprentice", "Intermediate", "Advanced", "Expert", "Master")
LESSONS: list[Lesson] = []

def _add(tier, prefix, rows):
    for i, row in enumerate(rows, 1):
        title, skill, mission, module = row
        LESSONS.append(Lesson(f"{prefix}-{i:02d}", tier, title, skill, mission, module))

_add("Novice", "N", [
    ("First Print", "print and expressions", "Make your pet greet the world.", "builtins"),
    ("Values", "numbers, strings, booleans, None", "Give your pet useful facts.", "builtins"),
    ("Variables", "assignment and names", "Track your pet's food and energy.", "builtins"),
    ("Operators", "arithmetic and comparisons", "Calculate supplies for a picnic.", "operator"),
    ("Input", "input and conversion", "Let your pet ask the player a question.", "builtins"),
    ("Decisions", "if, elif, else", "Teach when to eat or play.", "builtins"),
    ("Loops", "for and while", "Plant a row of flowers automatically.", "builtins"),
    ("Functions", "def, parameters, return", "Build a reusable care action.", "builtins"),
    ("Debugging", "tracebacks and experiments", "Find and repair a broken world action.", "traceback"),
])
_add("Apprentice", "A", [
    ("Collections", "list, tuple, set, dict", "Build a pet inventory.", "collections"),
    ("Comprehensions", "list/set/dict comprehensions", "Transform a garden plan.", "builtins"),
    ("Strings", "formatting and text processing", "Create a pet journal.", "string"),
    ("Files", "reading and writing files", "Save a world blueprint.", "pathlib"),
    ("Paths", "filesystem paths", "Organize world resources.", "pathlib"),
    ("Exceptions", "try/except/else/finally", "Make a safe market transaction.", "builtins"),
    ("Modules", "import and packages", "Turn a solution into a reusable tool.", "importlib"),
    ("Testing", "assertions and test cases", "Prove a pet-care function works.", "unittest"),
    ("Data", "JSON and CSV", "Export and restore world data.", "json"),
])
_add("Intermediate", "I", [
    ("Classes", "objects, classes, methods", "Create a custom world object.", "builtins"),
    ("Inheritance", "base classes and overriding", "Create specialized pet species from a shared class.", "builtins"),
    ("Composition", "objects containing objects", "Build a town from reusable world components.", "builtins"),
    ("Properties", "property getters and setters", "Protect and validate pet state.", "builtins"),
    ("Python Data Model", "dunder methods and protocols", "Teach a world object to behave naturally with Python operators.", "builtins"),
    ("Dataclasses", "structured state", "Model a pet and inventory.", "dataclasses"),
    ("Iterators", "iter and next", "Build a world resource stream.", "itertools"),
    ("Generators", "yield and lazy computation", "Generate an endless supply trail.", "builtins"),
    ("Decorators", "wrapping behavior", "Add rewards to world actions.", "functools"),
    ("Context Managers", "with and cleanup", "Safely manage a world resource.", "contextlib"),
    ("Typing", "type hints and protocols", "Make a reusable typed tool.", "typing"),
    ("Regex", "pattern matching", "Search and validate pet notes.", "re"),
    ("SQLite", "relational data", "Build a persistent market database.", "sqlite3"),
    ("CLI", "command-line interfaces", "Turn a world tool into a command.", "argparse"),
])
_add("Advanced", "V", [
    ("Dates", "datetime and time zones", "Schedule world events.", "datetime"),
    ("Math", "math, statistics, decimal, fractions", "Balance a world economy.", "math"),
    ("Randomness", "random and secrets", "Create fair encounters and secure tokens.", "random"),
    ("Concurrency", "threading and queues", "Run independent world jobs.", "threading"),
    ("Processes", "multiprocessing", "Parallelize a heavy simulation.", "multiprocessing"),
    ("Futures", "concurrent.futures", "Coordinate multiple tasks.", "concurrent.futures"),
    ("Async", "async/await", "Animate several world systems together.", "asyncio"),
    ("Networking", "sockets and protocols", "Make two world stations communicate.", "socket"),
    ("HTTP", "clients and servers", "Consume a safe web service.", "http"),
    ("Subprocesses", "process control", "Build a controlled automation tool.", "subprocess"),
])
_add("Expert", "E", [
    ("Tkinter", "desktop GUI", "Build a control panel for the pet.", "tkinter"),
    ("Tkinter Canvas", "2D interactive graphics", "Draw an interactive mini-world.", "tkinter"),
    ("Pygame", "game loop and input", "Build a playable pet mini-game.", "pygame"),
    ("Pygame Graphics", "sprites, surfaces, animation", "Animate your pet in a game scene.", "pygame"),
    ("Pygame Physics", "collision and timing", "Create a physics-based challenge.", "pygame"),
    ("Packaging", "build and distribute projects", "Package a world tool cleanly.", "importlib"),
    ("Logging", "diagnostics and structured logs", "Diagnose a failing world system.", "logging"),
    ("Security", "hashing, secrets, safe data", "Protect a world account token.", "hashlib"),
    ("Performance", "profiling and optimization", "Speed up a slow simulation.", "cProfile"),
    ("Architecture", "separation of concerns", "Refactor a growing world system.", "dataclasses"),
])
_add("Master", "M", [
    ("Standard Library Quest", "module selection", "Solve a world problem using the right standard-library tools.", "stdlib"),
    ("Full GUI Project", "Tkinter application design", "Build a complete world editor.", "tkinter"),
    ("Full Game Project", "Pygame architecture", "Build a complete playable game.", "pygame"),
    ("Automation Project", "integration and concurrency", "Automate a multi-stage world operation.", "stdlib"),
    ("Data Project", "database and data processing", "Build persistent world analytics.", "sqlite3"),
    ("Network Project", "networked application", "Build a robust client/server world feature.", "socket"),
    ("Testing Project", "quality engineering", "Test and repair a deliberately broken project.", "unittest"),
    ("Capstone", "independent Python engineering", "Design, build, debug, test, and explain a substantial world system.", "python"),
])

STANDARD_LIBRARY = [
    "builtins", "collections", "itertools", "functools", "operator", "math", "statistics",
    "decimal", "fractions", "random", "datetime", "zoneinfo", "calendar", "time", "pathlib",
    "os", "shutil", "glob", "tempfile", "fileinput", "json", "csv", "configparser", "sqlite3",
    "re", "string", "textwrap", "difflib", "enum", "dataclasses", "typing", "abc", "contextlib",
    "logging", "traceback", "warnings", "unittest", "doctest", "argparse", "subprocess", "threading",
    "multiprocessing", "concurrent.futures", "asyncio", "queue", "socket", "http", "urllib", "email",
    "xml", "html", "hashlib", "hmac", "secrets", "base64", "struct", "pickle", "shelve", "cProfile",
]
ECOSYSTEM = ["tkinter", "pygame"]

def curriculum() -> list[dict[str, Any]]:
    return [asdict(x) for x in LESSONS]

def lesson_for(index: int) -> dict[str, Any]:
    return asdict(LESSONS[index % len(LESSONS)])

def total_lessons() -> int:
    return len(LESSONS)
